### PR TITLE
Add code to use RabbitMQ, Swift, and SMTP email

### PR DESCRIPTION
**Pull request by fozboz**
# What does this Pull Request do?
We provide the implementation of the configuration changes made in duracloud-db and management-console.

We have added the ability to:
- Read audit logs from a Swift container
- Write audit events to a RabbitMQ queue
- Read/write notification events from/to a RabbitMQ queue
- Send emails using an SMTP server

# How should this be tested?

### Prerequisites for testing
- An SMTP server
- A RabbitMQ server with:
  - `rabbitmq_message_timestamp` plugin installed
  - a fanout exchange
  - a direct exchange
- A Swift endpoint running S3 emulation middleware layer

### Steps
You will need to build this repo first, then build management-console.

To test SMTP email, add this to the .properties file:
```
#Email Type: 'ses' or 'spring'
notification.type=spring
# Connection details for email notifications
notification.user=smtp_or_aws_username
notification.pass=smtp_or_aws_password
notification.from-address=someone@example.com
notification.admin-address=admin@example.com
#If the notification type is 'spring' (must be SMTP server)
notification.host=smtp.example.com
notification.port=25
```
then restart duradmin/durastore. The debug log should show a successful email connection test.

To test RabbitMQ notifications:
- Create a "fanout" exchange in RabbitMQ
- configure Global Properties in the AMA with RabbitMQ
- use the name of the fanout exchange as the "RabbitMQ exchange" option
- add a new space
- you should see successful message dispatch, receipt and deletion messages in the debug log
- you should also see that two queues were automatically created in RabbitMQ and bound to the fanout exchange

To test RabbitMQ audit queue:
- Create a "direct" exchange
- Create a queue and give it a {queuename}. Bind this queue to the direct exchange and use {queuename} as the routing key
- configure Duracloud Mill in the AMA with RabbitMQ
  - use the name of the direct exchange as the RabbitMQ Exchange
  - Use {queuename} as the Audit Queue
- Perform some audit-able actions. You should see messages appearing on the audit queue

To test audit logs:
- Add these parameters to the .properties file
```
swift.accessKey=accesskey
swift.secretKey=secretkey
swift.endpoint=https://s3.example.com
swift.signerType=signerType_if_required
```
- Restart duradmin/durastore
- Generate an audit log

# Notable changes
We have changed the package `package org.duracloud.common.sns` to `package org.duracloud.common.changenotifier` to more accurately reflect its purpose.

# Interested parties
@bbranan

### CODE REVIEW

File: common/src/main/java/org/duracloud/common/util/DuracloudConfigBean.java [hunk 1]
```diff
@@ -58,4 +58,24 @@ public String getAmaUrl() {
...
+
+    public String getSwiftSecretKey() {
+        return env.getProperty("swift.secretKey");
+    }
+
+    public String getSwiftEndpoint() {
+        return env.getProperty("swift.endpoint");
+    }
+
+    public String getSwiftSignerType() {
```
Comment (bbranan) [inline common/src/main/java/org/duracloud/common/util/DuracloudConfigBean.java]: If Signer Type is no longer being used, it should be removed.

File: notification-spring/src/test/java/org/duracloud/notification/SpringNotificationFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ */
+package org.duracloud.notification;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * @author Andrew Woods
+ * Date: 3/11/11
```
Comment (bbranan) [inline notification-spring/src/test/java/org/duracloud/notification/SpringNotificationFactoryTest.java]: Author and date here are probably not right

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
+package org.duracloud.common.changenotifier;
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java]: Rectified !880fba5

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+import com.rabbitmq.client.ConnectionFactory;
+import com.rabbitmq.client.DefaultConsumer;
+import com.rabbitmq.client.Envelope;
+import com.rabbitmq.client.ShutdownSignalException;
+import org.duracloud.common.error.DuraCloudRuntimeException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Shibo Liu
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java]: Rectified !880fba5

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/SubscriptionManager.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package org.duracloud.common.changenotifier;
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/SubscriptionManager.java]: Rectified !880fba5

File: notification-spring/src/test/java/org/duracloud/notification/SpringNotificationFactoryTest.java (hunk 2 above)
Comment (fozboz) [reply notification-spring/src/test/java/org/duracloud/notification/SpringNotificationFactoryTest.java]: Rectified !880fba5

File: notification-spring/src/main/java/org/duracloud/notification/SpringNotificationFactory.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    private Integer port;
+
+    public SpringNotificationFactory(String host, Integer port) {
+        this.host = host;
+        this.port = port;
+    }
+
+    @Override
+    public void initialize(String username, String password) {
+        String logUserMsg = "{}";
```
Comment (bbranan) [inline notification-spring/src/main/java/org/duracloud/notification/SpringNotificationFactory.java]: It's not clear why this defaults to `"{}"`. If username is null or empty, you want this to be an empty string, right? Why not just set it to `""`? As it is, in the case where username is null (see next comment), the log messages may look odd.

File: notification-spring/src/main/java/org/duracloud/notification/SpringNotificationFactory.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                logUserMsg + ", because {}", host, port,
+                username, ex.getMessage());
+        }
+    }
+
+    @Override
+    public Emailer getEmailer(String fromAddress) {
+        if (null == fromAddress ||
+            !EmailValidator.getInstance().isValid(fromAddress)) {
+            String msg = "fromAddress not valid notification: " + fromAddress;
```
Comment (bbranan) [inline notification-spring/src/main/java/org/duracloud/notification/SpringNotificationFactory.java]:
```suggestion
            String msg = "fromAddress " + fromAddress + " is not valid. Email notification service NOT initialized.";
```

File: notification-spring/src/main/java/org/duracloud/notification/SpringNotificationFactory.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            log.error(msg);
+            throw new IllegalArgumentException(msg);
+        }
+
+        if (null == emailService) {
+            String msg = "Emailer service !initialized.";
```
Comment (bbranan) [inline notification-spring/src/main/java/org/duracloud/notification/SpringNotificationFactory.java]:
```suggestion
            String msg = "emailService is null. Email notification service NOT initialized.";
```

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/SubscriptionManager.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ *     http://duracloud.org/license/
+ */
+package org.duracloud.common.changenotifier;
+
+/**
+ * @author Shibo Liu
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/SubscriptionManager.java]: Fixed in !d92bb28

File: notification-spring/pom.xml [hunk 10]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.duracloud</groupId>
+  <artifactId>notification-spring</artifactId>
```
[the remark this answers is not in the document]
Comment (fozboz) [reply notification-spring/pom.xml]: Classes and modules refactored in !f427f85

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+                    @Override
+                    public void handleConsumeOk(String consumerTag) {
+                        log.info("Consumer registered: {}", consumerTag);
+                        initialized = true;
+                    }
+
+                    @Override
+                    public void handleCancel(String consumerTag) {
+                        // consumer has been cancelled unexpectedly
+                        log.debug("Consumer has been cancelled unexpectedly: " + consumerTag);
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java]: This is called when the consumer is cancelled for reasons other than by a call to Channel.basicCancel. For example, the queue has been deleted. If there is a connection/server issue, `handleShutdownSignal` is called. I agree this should be a warn.

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+                       // either the channel or the underlying connection has been shut down.
+                        log.debug("Either the channel or the underlying connection has been shut down" +
+                                  " for consumer {} because {}", consumerTag, sig.getReason().toString());
+                        initialized = false;
+                    }
+
+                });
+
+        } catch (Exception e) {
+            log.debug("Consumer failed to subscribe: " + e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java]: Yeah I think most of these debugs are actually errors. Will change.

File: common-queue/src/main/java/org/duracloud/common/queue/rabbitmq/RabbitMQTaskQueue.java [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+            throw new DuraCloudRuntimeException(ex);
+        }
+    }
+
+    public RabbitMQTaskQueue(Connection conn, String exchange, String queueName) {
+        try {
+            this.exchangeName = exchange;
+            mqChannel = conn.createChannel();
+            mqChannel.queueBind(queueName, exchange, queueName);
+            queueUrl = "RabbitMQ-" + conn.getAddress();
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-queue/src/main/java/org/duracloud/common/queue/rabbitmq/RabbitMQTaskQueue.java]: The “RabbitMQ-” prefix is just to make the log message clearer, it doesn’t serve any programmatic purpose.

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java]: Fixed in !b4b21b0

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/RabbitMQSubscriptionManager.java]: Fixed in !b4b21b0

File: common-changenotifier/src/main/java/org/duracloud/common/changenotifier/config/SubscriptionManagerConfig.java [hunk 14]
```diff
@@ -37,22 +39,37 @@
...
             String queueName = "node-queue-" + appName + "-" +
                                Inet4Address.getLocalHost().getHostName().replace(".", "_");
-            SnsSubscriptionManager subscriptionManager =
-                new SnsSubscriptionManager(AmazonSQSClientBuilder.defaultClient(),
-                                           AmazonSNSClientBuilder.defaultClient(),
-                                           props.getInstanceNotificationTopicArn(), queueName);
+
+            SubscriptionManager subscriptionManager;
+
+            if (props.getNotifierType().equalsIgnoreCase("RABBITMQ")) {
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-changenotifier/src/main/java/org/duracloud/common/changenotifier/config/SubscriptionManagerConfig.java]: Fixed in !dc82244

File: durastore/src/main/java/org/duracloud/durastore/util/StorageProviderFactoryImpl.java [hunk 15]
```diff
@@ -137,7 +138,22 @@ private void configureAuditQueue(AuditConfig auditConfig) {
                 // If no queue name is defined, turn off auditing
                 this.auditQueue = new NoopTaskQueue();
             } else {
-                this.auditQueue = new SQSTaskQueue(queueName);
+                String queueType = auditConfig.getAuditQueueType();
+                if (queueType.equalsIgnoreCase("RABBITMQ")) {
```
[the remark this answers is not in the document]
Comment (fozboz) [reply durastore/src/main/java/org/duracloud/durastore/util/StorageProviderFactoryImpl.java]: Fixed in !dc82244

File: common-queue/src/main/java/org/duracloud/common/queue/rabbitmq/RabbitMQTaskQueue.java [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+                    String exchange = envelope.getExchange();
+                    long deliveryTag = envelope.getDeliveryTag();
+                    Long sentTime = properties.getTimestamp().getTime();
+                    Long preworkQueueTime = System.currentTimeMillis() - sentTime;
+                    log.info(
+                        "RabbitMQ message received - queue: {}, queueUrl: {}, deliveryTag: {}, preworkQueueTime: {}"
+                        , queueName, queueUrl, deliveryTag
+                        , DurationFormatUtils.formatDuration(preworkQueueTime, "HH:mm:ss,SSS"));
+                    Task task = marshallTask(body, deliveryTag, routingKey, exchange);
+                    task.setVisibilityTimeout(visibilityTimeout);
```
[the remark this answers is not in the document]
Comment (fozboz) [reply common-queue/src/main/java/org/duracloud/common/queue/rabbitmq/RabbitMQTaskQueue.java]: It doesn't hurt to set it, but yes, RabbitMQ doesn't use this, so removed in !dc82244

File: common-queue/src/main/java/org/duracloud/common/queue/rabbitmq/RabbitMQTaskQueue.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (bbranan) [reply common-queue/src/main/java/org/duracloud/common/queue/rabbitmq/RabbitMQTaskQueue.java]: Oh. I'd be concerned about that causing confusion, as the log message suggests that the prefix is part of the URL. Perhaps something like `"(RabbitMQ) "` would be more clearly a prefix?

**Pull request merged**